clang/lib/Sema/SemaTemplate.cpp

```diff
@@ -3058,6 +3058,133 @@ void Sema::NoteAllFoundTemplates(TemplateName Name) {
   }
 }
 
+static QualType commonTypeImpl(Sema &S, TemplateName BaseTemplate,
+                               SourceLocation TemplateLoc,
+                               ArrayRef<TemplateArgument> Ts) {
+  auto lookUpCommonType = [&](TemplateArgument T1,
+                              TemplateArgument T2) -> QualType {
+    // Don't bother looking for other specializations if both types are
+    // builtins - users aren't allowed to specialize for them
```

Comment:
Can we diagnose the case where users do this, please? Similarly for specialising `common_type<>`, `common_type<T>`, and `common_type<T1, T2, T3, Ts...>`.

Reply:
We can do that, but I don't think that's part of this project. AFAICT this requires tinkering with the template specialization code and check whether a specialization is allowed to exist. To me that seems like an entire project on its own.

Reply:
We need to handle the first case in this patch because it is currently silently changing the outcome of `common_type<builtin, builtin>`. Although the standard gives us licence to silently change this, it isn't fair to users who either weren't aware of the consequences of specialising this, or are depending on such code.

Reply:
How is this any different from all the other cases where we switched to builtins, which results in potential breaks?

Reply:
Good point. There are only a handful of cases that this applies to, but they should also break loudly. In that case, I'd prefer it if getting that diagnostic logic into Clang preceded submitting this PR.

Reply:
Given the precedent I don't think it's reasonable to block this patch on diagnosing some edge cases.

Reply:
It is only precedent due to it not having been previously recognised as an issue. It has now been identified and should be addressed, instead of increasing the problem's surface area.

We should do things right, not wrong because "they were done that way previously".

Reply:
I believe it would make sense for such a diagnostic to work today, even without the builtin. For example, it would be nice for the following code to be diagnosed with a warning (or an error, whatever):

```c++
#include <type_traits>

template <>
struct std::common_type<int, int> { using type = long; }; // should warn

struct Foo { };
template <>
struct std::common_type<Foo> { using type = int; }; // should warn

// etc...
```

I don't think this should be tied to the implementation of the builtin, i.e. it should be diagnosed where the user writes the invalid specialization, not where we use `std::common_type_t`. As such, I think it would make sense to tackle this as a follow-up and I filed an issue for it: https://github.com/llvm/llvm-project/issues/101509